### ISSUE TITLE
Trackmania 2020, Trackmania 2, Trackmania Nations Forever

### ISSUE THREAD

**Issue by Nicklander** [parkervcp/eggs#719]
---------------------------------------------------------------------------------------------------------------

Service: Trackmania 2020 + Pyplanet Server Controller

Does this expand an already existing service: N

Link to game: 
https://www.trackmania.com/

Links for server downloads:
TM2020 Server: http://files.v04.maniaplanet.com/server/TrackmaniaServer_Latest.zip
Pyplanet: https://pypla.net/en/latest/

Links for install steps/docs:
TM2020: https://forums.ubisoft.com/showthread.php/2242192-Tutorial-Trackmania-2020-Dedicated-Server
Pyplanet: https://pypla.net/en/latest/intro/index.html

Note: The server controllers also have to be installed to be able to control the server properly, otherwise people won't use it...

---------------------------------------------------------------------------------------------------------------

Service: Trackmania 2 (Maniaplanet) + Maniacontrol Server Controller

Link to game: 
https://www.maniaplanet.com/

Links for server downloads:
TM2 Server: http://files.v04.maniaplanet.com/server/ManiaplanetServer_Latest.zip
Maniacontrol: https://www.maniacontrol.com/download

Links for install steps/docs:
TM2 (Maniaplanet): https://doc.maniaplanet.com/dedicated-server/getting-started
Maniacontrol: https://github.com/ManiaControl/ManiaControl#setup

---------------------------------------------------------------------------------------------------------------

Service: Trackmania Nations Forever (TMNF) + Xaseco Server Controller

Link to game: 
http://trackmaniaforever.com/

Links for server downloads:
TMNF Server: http://files2.trackmaniaforever.com/TrackmaniaServer_2011-02-21.zip
Xaseco: https://www.xaseco.org/overview.php#download

Links for install steps/docs: Check attached PDF
[TMNF Xaseco Quickstart Guide.pdf](https://github.com/parkervcp/eggs/files/5324463/TMNF.Xaseco.Quickstart.Guide.pdf)

---------------------------------------------------------------------------------------------------------------

**Comment by DatMayo** [comment]
Pre-Build of your requested "Trackmainia 2020" script, please check and report any errors, so I can correct them
[egg-trackmania2020.zip](https://github.com/parkervcp/eggs/files/5761932/egg-trackmania2020.zip)

**Comment by treier57lol** [comment]
WIP TMNF/UF Server: https://github.com/Hoerli1337/pterodactyl-eggs
Standalone server fully functional, XASECO coming soon!

**Comment by IIPoliII** [comment]
Looks awesome! Looking forward for it.

**Comment by IIPoliII** [comment]
I can't make the XML parser work at all. I don't understand how it works but it really truly does shit.

**Comment by Nicklander** [comment]
> Pre-Build of your requested "Trackmainia 2020" script, please check and report any errors, so I can correct them
> [egg-trackmania2020.zip](https://github.com/parkervcp/eggs/files/5761932/egg-trackmania2020.zip)

Thank you, but I would need a TM2020 + Pyplanet setup. A standalone TM2020 server is quite pointless for me.

**Comment by Nicklander** [comment]
> WIP TMNF/UF Server: https://github.com/Hoerli1337/pterodactyl-eggs
> Standalone server fully functional, XASECO coming soon!
Thank you, I will try out the TM2 + Pyplanet & TMNF + Xaseco eggs next week.

**Comment by treier57lol** [comment]
Forgot to mention my TM2 + PyPlanet egg here: https://github.com/treier57lol/docker-trackmania/tree/pyplanet
Could still use some final tweaks but it works, been using it for over a month and it runs great. Converting it to a TM2020 + PyPlanet egg should be pretty easy.

**Comment by Nicklander** [comment]
> Forgot to mention my TM2 + PyPlanet egg here: https://github.com/treier57lol/docker-trackmania/tree/pyplanet
> Could still use some final tweaks but it works, been using it for over a month and it runs great. Converting it to a TM2020 + PyPlanet egg should be pretty easy.

Yes that would be great !
I'll have more time to try all of that out mid-march.

**Comment by Nicklander** [comment]
> 
> 
> Forgot to mention my TM2 + PyPlanet egg here: https://github.com/treier57lol/docker-trackmania/tree/pyplanet
> Could still use some final tweaks but it works, been using it for over a month and it runs great. Converting it to a TM2020 + PyPlanet egg should be pretty easy.

Did you manage to convert it to a TM2020 + PyPlanet egg ? I would still need that one.

**Comment by Jowcey** [comment]
> > Forgot to mention my TM2 + PyPlanet egg here: https://github.com/treier57lol/docker-trackmania/tree/pyplanet
> > Could still use some final tweaks but it works, been using it for over a month and it runs great. Converting it to a TM2020 + PyPlanet egg should be pretty easy.
> 
> Did you manage to convert it to a TM2020 + PyPlanet egg ? I would still need that one.

Was this ever completed?

**Comment by w00p** [comment]
I would love to see this get completed!

**Comment by IIPoliII** [comment]
If I remember correctly I made it to work :
https://gist.github.com/IIPoliII/c34874836aec845619a3df76b48ad002

But I can't assure it



> I would love to see this get completed!

**Comment by w00p** [comment]
Yes, i did test that in fact. But Trackmania client do not show the server on local network. 
I cant find a way to connect/add it in the client by ip.
I have Maniaplanet and Pyplanet setup a ubuntu server, and that one get recognized.

**Comment by w00p** [comment]
Yeah, but it works. And the Trackmania server I starting. But I can't get Trackmania Client to show it.

I read on a form something like this:
Trackmania 2020's server doesn't broadcast anything. To find existing servers a Trackmania 2020 client broadcasts a UDP packet to a link-local network (I don't know how it is picked) to ports 2350 to 2400. Then it establishes a TCP connection to all hosts who replied to the ports with which the host replied. Then it probably gets server information and displays it on the server-list in the game.

And yes, I have allocated port 2350 to my node and server.

**Comment by IIPoliII** [comment]
> > If I remember correctly I made it to work : https://gist.github.com/IIPoliII/c34874836aec845619a3df76b48ad002
> > But I can't assure it
> > > I would love to see this get completed!
> 
> That needs definitly a update to the image and install image

Yeah it's definitily old I did it a moment ago with pieces I found around
